### DISCUSSION TITLE
Is it possible to add device Tuya TH05Z

### DISCUSSION THREAD

**Discussion by Rincovanderbeek** [Koenkk/zigbee2mqtt#19702] [Devices]
Can anyone help me adding this device:  Tuya temperature & humidity sensor TH05Z
Currently zigbee2mqtt sees this device as a TS0601  and then start with enless errors.

https://nl.aliexpress.com/item/1005005897159196.html?spm=a2g0o.productlist.main.21.75b81830r75oCM&algo_pvid=5a02cfbf-9ac8-4c96-a5d0-9e5dd7d3e664&algo_exp_id=5a02cfbf-9ac8-4c96-a5d0-9e5dd7d3e664-10&pdp_npi=4%40dis%21EUR%2126.88%219.68%21%21%2128.05%21%21%40211b81a316999062114057606e93e0%2112000034755832541%21sea%21NL%21828594121%21&curPageLogUid=vogxsBQL8LsA

**Comment by pubFbx** [comment]
Hello, I confirm the problem and I don't have the know-how to make a 
supported device

Warning 2023-11-18 18:51:32Received message from unsupported device with Zigbee model 'TS0601' and manufacturer name '_TZE200_vvmbj46n'

**Comment by sweepy-77** [comment]
Hello look here
[https://github.com/Koenkk/zigbee2mqtt/issues/19731]

**Reply by pubFbx** [reply]
Hello, thank you for this information, but I don't have a sufficient level of competence to integrate this code into Z2M. I'll wait and hope that the next version will include this device.

**Reply by Rincovanderbeek** [reply]
Same counts for me

**Comment by Koenkk** [comment]
what is the data/database.db entry of this device?

**Reply by sweepy-77** [reply]
{"id":12,"type":"EndDevice","ieeeAddr":"0xa4c1384eb0d22c4f","nwkAddr":65187,"manufId":4417,"manufName":"_TZE200_vvmbj46n","powerSource":"Battery","modelId":"TS0601","epList":[1],"endpoints":{"1":{"profId":260,"epId":1,"devId":81,"inClusterList":[4,5,61184,0],"outClusterList":[25,10],"clusters":{"genBasic":{"attributes":{"65503":"d#�,id#�,fe#�,ie#�,\u0012","65506":56,"65508":0,"stackVersion":0,"dateCode":"","appVersion":72}}},"binds":[{"cluster":0,"type":"endpoint","deviceIeeeAddress":"0x00124b0022988f8b","endpointID":1}],"configuredReportings":[],"meta":{}}},"appVersion":72,"stackVersion":0,"hwVersion":1,"dateCode":"","zclVersion":3,"interviewCompleted":true,"meta":{"configured":821693351},"lastSeen":1700506744908,"defaultSendRequestWhen":"immediate"}

**Reply by Koenkk** [reply]
Could you check if the issue is fixed with the following external converter: https://gist.github.com/Koenkk/78ac831ec1143edc92e8231eed73ebd6

- save this as file next to `configuration.yaml` as `ext_converter.js`
- add it to `configuration.yaml`:
```yaml
external_converters:
  - ext_converter.js
```
- start z2m, check if issue is fixed

**Reply by sweepy-77** [reply]
Hello Koenkk,
This log:

```Zigbee2MQTT:debug 2023-11-22 11:40:00: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0],"type":"Buffer"},"datatype":4,"dp":9}],"seq":19456}' from endpoint 1 with groupID 0
Zigbee2MQTT:info  2023-11-22 11:40:00: MQTT publish: topic 'zigbee2mqtt/0xa4c1384eb0d22c4f', payload '{"battery_state":null,"humidity":45,"linkquality":160,"temperature":21,"temperature_unit":"celsius"}'
Zigbee2MQTT:debug 2023-11-22 11:40:00: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,1,134],"type":"Buffer"},"datatype":2,"dp":10}],"seq":19712}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:00: Datapoint 10 not defined for '_TZE200_vvmbj46n' with value 390
Zigbee2MQTT:debug 2023-11-22 11:40:00: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,0],"type":"Buffer"},"datatype":2,"dp":11}],"seq":19968}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:00: Datapoint 11 not defined for '_TZE200_vvmbj46n' with value 0
Zigbee2MQTT:debug 2023-11-22 11:40:00: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,60],"type":"Buffer"},"datatype":2,"dp":12}],"seq":20224}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:00: Datapoint 12 not defined for '_TZE200_vvmbj46n' with value 60
Zigbee2MQTT:debug 2023-11-22 11:40:00: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,20],"type":"Buffer"},"datatype":2,"dp":13}],"seq":20480}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:00: Datapoint 13 not defined for '_TZE200_vvmbj46n' with value 20
Zigbee2MQTT:debug 2023-11-22 11:40:01: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,120],"type":"Buffer"},"datatype":2,"dp":17}],"seq":20736}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:01: Datapoint 17 not defined for '_TZE200_vvmbj46n' with value 120
Zigbee2MQTT:debug 2023-11-22 11:40:01: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,120],"type":"Buffer"},"datatype":2,"dp":18}],"seq":20992}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:01: Datapoint 18 not defined for '_TZE200_vvmbj46n' with value 120
Zigbee2MQTT:debug 2023-11-22 11:40:01: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,6],"type":"Buffer"},"datatype":2,"dp":19}],"seq":21248}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:01: Datapoint 19 not defined for '_TZE200_vvmbj46n' with value 6
Zigbee2MQTT:debug 2023-11-22 11:40:01: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandDataReport', cluster 'manuSpecificTuya', data '{"dpValues":[{"data":{"data":[0,0,0,6],"type":"Buffer"},"datatype":2,"dp":20}],"seq":21504}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:01: Datapoint 20 not defined for '_TZE200_vvmbj46n' with value 6
Zigbee2MQTT:debug 2023-11-22 11:40:01: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":85}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:01: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":85}'
Zigbee2MQTT:debug 2023-11-22 11:40:01: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":87}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:01: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":87}'
Zigbee2MQTT:debug 2023-11-22 11:40:03: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":92}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:03: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":92}'
Zigbee2MQTT:debug 2023-11-22 11:40:05: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":97}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:05: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":97}'
Zigbee2MQTT:debug 2023-11-22 11:40:07: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":102}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:07: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":102}'
Zigbee2MQTT:debug 2023-11-22 11:40:08: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":106}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:08: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":106}'
Zigbee2MQTT:debug 2023-11-22 11:40:10: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuSyncTime', cluster 'manuSpecificTuya', data '{"payloadSize":110}' from endpoint 1 with groupID 0
Zigbee2MQTT:debug 2023-11-22 11:40:10: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuSyncTime' and data '{"payloadSize":110}'`

and database: 

`{"id":12,"type":"EndDevice","ieeeAddr":"0xa4c1384eb0d22c4f","nwkAddr":59659,"manufId":4417,"manufName":"_TZE200_vvmbj46n","powerSource":"Battery","modelId":"TS0601","epList":[1],"endpoints":{"1":{"profId":260,"epId":1,"devId":81,"inClusterList":[4,5,61184,0],"outClusterList":[25,10],"clusters":{"genBasic":{"attributes":{"65503":"���,i���,i���,i���,i���,i���,f���,\u0012","65506":56,"65508":0,"65534":0,"stackVersion":0,"dateCode":"","manufacturerName":"_TZE200_vvmbj46n","zclVersion":3,"appVersion":72,"modelId":"TS0601","powerSource":3}}},"binds":[],"configuredReportings":[],"meta":{}}},"appVersion":72,"stackVersion":0,"hwVersion":1,"dateCode":"","zclVersion":3,"interviewCompleted":true,"meta":{"configured":1335756102},"lastSeen":1700648675696,"defaultSendRequestWhen":"immediate"}
`

**Reply by Koenkk** [reply]
Thanks for the log, but does the device work now?

**Reply by sweepy-77** [reply]
battery_state does not return any null value info.
some datapoints are not recognized.
I found that it worked better with the model code: JM-TRH-ZGB-V1 set external_converters

**Reply by Koenkk** [reply]
To find out what all these datapoints mean, see https://www.zigbee2mqtt.io/advanced/support-new-devices/03_find_tuya_data_points.html

**Reply by sweepy-77** [reply]
I don't have a tuya zigbee gateway. I couldn't test it right away.

EDIT:
After tests
battery state is not known I removed it.
I put `e.battery() and datapoints [4, 'battery', tuya.valueConverter.raw],` the battery value goes up nicely.

**Reply by Koenkk** [reply]
Great! Are the other values also there?

**Reply by sweepy-77** [reply]
Below:
`   

    exposes: [e.temperature(), e.humidity(), e.battery(), tuya.exposes.temperatureUnit(),],
    meta: {
        tuyaDatapoints: [
            [1, 'temperature', tuya.valueConverter.divideBy10],
            [2, 'humidity', tuya.valueConverter.raw],
            [4, 'battery', tuya.valueConverter.raw],
            [9, 'temperature_unit', tuya.valueConverter.temperatureUnitEnum],
            [10, 'max_temp', tuya.valueConverter.divideBy10],
            [11, 'min_temp', tuya.valueConverter.divideBy10],
            [12, 'max_humidity', tuya.valueConverter.raw],
            [13, 'min_humidity', tuya.valueConverter.raw],
            [14, 'alarm_temp', tuya.valueConverter.trueFalse1],
            [15, 'alarm_humidity', tuya.valueConverter.trueFalse1],
            [17, 'interval_report_temp', tuya.valueConverter.raw],
            [18, 'interval_report_humidity', tuya.valueConverter.raw],
            [19, 'sensitivity_temp', tuya.valueConverter.raw],
            [20, 'sensitivity_humidity', tuya.valueConverter.raw],
        ],`
tuyaDatapoints 14 and 15 only appear if an alarm is triggered.

**Reply by Koenkk** [reply]
Cool, would you mind making a PR to add support for this device?

**Reply by Rincovanderbeek** [reply]
I'm back :-)
I see a lot of work already has been done, thanks!
As I am pretty new, what is a PR?  (Personal Reply ?)

**Reply by Rincovanderbeek** [reply]
I think I see, Pull Request

**Reply by Rincovanderbeek** [reply]
mmm, not an option, I got 'Pull request creation failed. Validation failed: must be a collaborator'

**Reply by sweepy-77** [reply]
Before creating the PR I need to look at how to set the min and max alarms (dp: 10, 11, 12, 13). and also the dp values: 14, 15 because the values are 0 for min, 1 for max and 2 for no alarm.

I also have this message that comes up often.

`

debug 2023-11-24 21:33:45: Received Zigbee message from '0xa4c1384eb0d22c4f', type 'commandMcuVersionResponse', cluster 'manuSpecificTuya', data '{"seq":35965,"version":202}' from endpoint 1 with groupID 0
debug 2023-11-24 21:33:45: No converter available for 'ZTH05Z' with cluster 'manuSpecificTuya' and type 'commandMcuVersionResponse' and data '{"seq":35965,"version":202}'`

**Reply by sweepy-77** [reply]
Hello Koenkk,

How to send these commands to the device.This is the command to modify the values seen in iot tuya platform

commands : 
[{code: "temp_unit_convert", value: "c"}, {code: "maxtemp_set", value: 250},…]
    0: {code: "temp_unit_convert", value: "c"}
     code:"temp_unit_convert"
     value: "c"
    1: {code: "maxtemp_set", value: 250}
     code: "maxtemp_set"
     value: 250
    2: {code: "minitemp_set", value: 160}
     code: "minitemp_set"
     value: 160
    3: {code: "maxhum_set", value: 70}
     code: "maxhum_set"
     value: 70
    4: {code: "minihum_set", value: 30}
     code: "minihum_set"
     value: 30
    5: {code: "temp_periodic_report", value: 120}
     code: "temp_periodic_report"
     value: 120
    6: {code: "hum_periodic_report", value: 120}
     code: "hum_periodic_report"
     value: 120
    7: {code: "temp_sensitivity", value: 6}
     code: "temp_sensitivity"
     value: 6
    8: {code: "hum_sensitivity", value: 6}
     code: "hum_sensitivity"
     value: 6

**Reply by Koenkk** [reply]
Probably these can all be mapped using `tuya.valueConverter.divideBy10` or `tuya.valueConverter.raw`, where do you get stuck with updating your external converter?

**Reply by sweepy-77** [reply]
example if I run the mqtt command
zigbee2mqtt/my_device/set {"min_temperature": 15} I have no error but nothing changes
et 
zigbee2mqtt/my_device/set {"minitemp_set": 15} I have the error : no converter available.

toZigbee.js must be missing something or how to put the command in the external converter

**Reply by Koenkk** [reply]
Add the missing keys by changing your ext converter like this: https://gist.github.com/Koenkk/78ac831ec1143edc92e8231eed73ebd6/revisions

**Reply by sweepy-77** [reply]
I made the modifications see here:
https://gist.github.com/sweepy-77/772f25eb2efca35689a5309377d0da93/revisions

The modifications work but are not remembered in the device.
THANKS

**Reply by Koenkk** [reply]
If I understand correctly, you change the `minitemp_set` but nothing changes on the device now? How did you verify this?

**Reply by sweepy-77** [reply]
I configure the device via the Tuya app ex: minitemp_set 25 I have an interior temperature of 21 so the temp_alarm alarm sends me lower alarm. When I put the device back on zigbee2mqtt I have the values set before minitemp_set to 25 and lower alarm. I modify minitemp_set via the cursor to 15 in the logs I can clearly see the command that was passed. if I press the button on the device it sends the data with the minitemp_set value at 25.

**Reply by sweepy-77** [reply]
Hi
I managed to send the data to the device.
In Iottuya I had seen that there were 2 commands to send, one empty and the other with the data. 
On the device there's a backlight button (also used for pairing, press for 5 seconds). When pressed (1 second), with the light on, you launch the command or move the cursors, and the changes are taken into account.

**Comment by Rincovanderbeek** [comment]
Thanks Koen,

I’m abroad till Thursday. Will check that by Friday, ok?

Met vriendelijke groet,

Rinco van der Beek

  *   verzonden vanaf een mobiel device

________________________________
Van: Koen Kanters ***@***.***>
Verzonden: Tuesday, November 21, 2023 8:29:59 PM
Aan: Koenkk/zigbee2mqtt ***@***.***>
CC: Rincovanderbeek ***@***.***>; Author ***@***.***>
Onderwerp: Re: [Koenkk/zigbee2mqtt] Is it possible to add device Tuya TH05Z (Discussion #19702)


Could you check if the issue is fixed with the following external converter: https://gist.github.com/Koenkk/78ac831ec1143edc92e8231eed73ebd6

  *   save this as file next to configuration.yaml as ext_converter.js
  *   add it to configuration.yaml:

external_converters:
  - ext_converter.js

  *   start z2m, check if issue is fixed

**Comment by jabss** [comment]
Hello, 

I've successfully added this device to my ZB2MQTT v1.35.1. Thanks!
It seems to be working fine (I'm just using the basic and default parameters: Current Temp and Hum, not messing with alarms or anything else).
Just have a question regarding this device: It has date and time, and these were "automagically" updated when adding the device to ZB2MQTT. Since these aren't in the "exposes" tab, can anyone elaborate how did it update date and time? Is zigbee protocol sending this info? If so, does it send to all devices? If not, how did it guessed it needed to send this info to this specific device?
I'm not complaining or anything, just intrigued about how this part worked.

Thanks!

**Comment by lutz108** [comment]
Hey!

First of all thanks to your great work for the community! I'm quite new to this topic, but (slowly) get how HA/zigbee/mqtt works :)

Now to my problem: I have two of those TuYa TH05Z sensors. They are ok (measuring temperature precisely is difficult), but I do get errors:

`Invalid value for number.thermometer_lcd_wohnzimmer_temperature_report_interval: 120 (range 5.0 - 60.0) `

I also can't change the value manually, neither in zigbee2mqtt nor in the mqtt device settings in HA. Whenever I change the value and revisit/reload the setting pages, it's back 120 (in zigbee2mqtt, probably root of error message?) and 35 in HA.


![image](https://github.com/Koenkk/zigbee2mqtt/assets/47532211/78b2e485-7135-4609-b2d2-2d985b9a47f1)

A corresponding mqtt log entry from zigbee2mqtt:
```MQTT publish: topic 'zigbee2mqtt/Thermometer LCD Wohnzimmer', payload '{"battery":131,"humidity":45,"humidity_alarm":"canceled","humidity_report_interval":120,"humidity_sensitivity":6,"linkquality":207,"max_humidity":60,"max_temperature":39,"min_humidity":20,"min_temperature":0,"temperature":21.8,"temperature_alarm":"canceled","temperature_report_interval":120,"temperature_sensitivity":0.6,"temperature_unit_convert":"celsius"}'```

Can this be an error by zigbee2mqtt or would you suspect another tool in the chain?


Side question: I also get `Failed to apply calibration to 'temperature': 'temperature_precision' is not a number, got string ()`. I use better-thermostat with temperature sensors (either those from this topic here or [ZG-227Z](https://www.zigbee2mqtt.io/devices/ZG-227Z.html#tuya-zg-227z) (in a lenght, not round version) and the [ME168](https://www.zigbee2mqtt.io/devices/ME168.html#avatto-me168) thermostates. Precision is set to 1. Is this more of a better-thermostat related issue?

**Reply by sweepy-77** [reply]
When pressed (1 second), with the light on, you launch the command or move the cursors, and the changes are taken into account.

**Reply by lutz108** [reply]
ouh sorry, I've over read your comment on that above. That worked, thanks a lot! :)

Since I'm new, I have two follow up question:
- Is that an odd, device specific behavior or something that one can expect in general from (tuya) zigbee devices (with zigbee2mqtt)?
- I assume the report is just the interval the device sends the current state by itself. But whenever some integration requests the temperature, i get the actual current value? Thus the report interval is not equal to the temporal resolution?

**Comment by Le-Ze** [comment]
Hello.
I bought 4 thermometers and I don't work properly.
I have Zigbee2mqtt@1.36.0 Build
You write that it is supported. Shows me tuya zth05z
https://www.zigbee2mqtt.io/devices/zth05z.html#tuya-zth05z
https://ibb.co/tqspwj6

The data goes wrong. Long silent. And then he sends a lot of data at once.
https://ibb.co/y2q3bxw
